motionplan/motionPlanner.go

```diff
@@ -488,80 +487,71 @@ func CheckPlan(
-		formerRunningPose := spatialmath.PoseBetweenInverse(lastPose, pathPosition)
-
-		// convert planNode's poses to be in absolute coordinates
-		if planNodes, err = rectifyTPspacePath(planNodes, sf, formerRunningPose); err != nil {
-			return err
-		}
+	// offset the plan using the errorState
+	offsetPlan := OffsetPlan(plan, errorState)
+	poses, err := offsetPlan.Path().GetFramePoses(checkFrame.Name())
+	if err != nil {
+		return err
 	}
 
-	// adjust planNodes by the errorState
-	// this only changes a node's pose and not its inputs
-	planNodes = transformNodes(planNodes, errorState)
-
-	// pre-pend node with current position of robot to planNodes
-	// Note that currentPosition is assumed to have accounted for the errorState
-	// Note that currentInputs is assumed to have NOT accounted for the errorState
-	planNodes = append([]node{&basicNode{pose: currentPosition, q: currentInputs}}, planNodes...)
-
-	// create constraints
+	// setup the planOpts
 	if sfPlanner.planOpts, err = sfPlanner.plannerSetupFromMoveRequest(
-		currentPosition,                    // starting pose
-		planNodes[len(planNodes)-1].Pose(), // goalPose
-		plan[0],                            // starting configuration
+		currentPose,
+		poses[len(poses)-1],
+		currentInputs,
 		worldState,
 		nil, // no pb.Constraints
 		nil, // no plannOpts
 	); err != nil {
 		return err
 	}
 
-	// go through plan and check that we can move from plan[i] to plan[i+1]
-	var totalTravelDistanceMM float64
-	for i := 0; i < len(planNodes)-1; i++ {
-		currentPose := planNodes[i].Pose()
-		nextPose := planNodes[i+1].Pose()
-		startConfiguration := planNodes[i].Q()
-		endConfiguration := planNodes[i+1].Q()
-
+	createSegment := func(
+		currPose, nextPose spatialmath.Pose,
+		currInput, nextInput map[string][]frame.Input,
+	) (*ik.Segment, error) {
+		currInputSlice, err := sf.mapToSlice(currInput)
+		if err != nil {
+			return nil, err
+		}
+		nextInputSlice, err := sf.mapToSlice(nextInput)
+		if err != nil {
+			return nil, err
+		}
 		// If we are working with a PTG plan we redefine the startConfiguration in terms of the endConfiguration.
 		// This allows us the properly interpolate along the same arc family and sub-arc within that family.
 		if relative {
-			startConfiguration = []frame.Input{
-				{Value: endConfiguration[0].Value}, {Value: endConfiguration[1].Value}, {Value: 0},
-			}
+			currInputSlice = []frame.Input{{Value: nextInputSlice[0].Value}, {Value: nextInputSlice[1].Value}, {Value: 0}}
 		}
-		segment := &ik.Segment{
-			StartPosition:      currentPose,
+		return &ik.Segment{
+			StartPosition:      currPose,
 			EndPosition:        nextPose,
-			StartConfiguration: startConfiguration,
-			EndConfiguration:   endConfiguration,
+			StartConfiguration: currInputSlice,
+			EndConfiguration:   nextInputSlice,
 			Frame:              sf,
+		}, nil
+	}
+
+	// create a list of segments to iterate through
+	segments := make([]*ik.Segment, 0, len(poses))
```

Comment:
I'm not convinced this case was ever working and therefore this isnt a regression.  Before we were building the path from the inputs but it was not possible to say where along the given segment you were. Now I want to be projecting the path from the current position outward, but we still have the same problem of not knowing where along the previous segment we were.  Ultimately I think we need to think about this more and it should not hold up this work which is otherwise a refactor

Reply:
I think Miko is correct here; what he's describing was formerly accounted for with this (deleted in the PR) code:
```

	if relative {
		// get pose of robot along the current trajectory it is executing
		lastPose, err := sf.Transform(currentInputs)
		if err != nil {
			return err
		}

		// where ought the robot be on the plan
		pathPosition := spatialmath.PoseBetweenInverse(errorState, currentPosition)

		// absolute pose of the previous node we've passed
		formerRunningPose := spatialmath.PoseBetweenInverse(lastPose, pathPosition)

		// convert planNode's poses to be in absolute coordinates
		if planNodes, err = rectifyTPspacePath(planNodes, sf, formerRunningPose); err != nil {
			return err
		}
```

We *do* have the ability to say where along a given segment we are, that is the `currentInputs` that is passed in.

Reply:
`currentInputs` is *not* the inputs of the currently executing segment. It is the amount of the current segment that has been executed.

If the plan that you started executing is (as above):
```
[ [0,0,0]
[ 0.2, 0.3, 200],
[ 0.4, 0.5, 600],
[0,0,0]]
```
and you are halfway through executing `[ 0.2, 0.3, 200]`, then your current inputs passed in will be `[ 0.2, 0.3, 100]`.

Reply:
Looking at the existing code more closely, you're right; we do not check this properly.

We do this:
```
	// pre-pend node with current position of robot to planNodes
	// Note that currentPosition is assumed to have accounted for the errorState
	// Note that currentInputs is assumed to have NOT accounted for the errorState
	planNodes = append([]node{&basicNode{pose: currentPosition, q: currentInputs}}, planNodes...)
```
But then as we create the segments we drop the distance from the current inputs, so we do not properly account for this.

However; now that we know this, we should correct this. `CheckPlan` will not operate properly if this is not done.

Reply:
thanks for jumping in @biotinker

Reply:
This ticket is super bloated already would it be ok to make this a fast follow and merge as is?

Reply:
hmmmm, I will defer to @biotinker's opinion on this.
Should we decide to defer this work, can you please create a ticket so that this bug fix is tracked?
If anything, I can likely take over this as part of RSDK-4511.

Reply:
I think it makes sense to roll into RSDK-4511.  Otherwise we could ticket it separately, I'm good with either

Reply:
+1 for rolling into RSDK-4511

Reply:
Personally I prefer not to merge something that we know is broken (even if it was already broken before). Since it's changing, the broken-ness could surface in new unexpected ways rather than the current stable way.

However it appears I am outvoted.